core/src/main/java/org/apache/calcite/rel/rules/AggregateUnionTransposeRule.java

```diff
@@ -151,9 +154,36 @@ public AggregateUnionTransposeRule(Class<? extends Aggregate> aggregateClass,
 
     // create a new union whose children are the aggregates created above
     relBuilder.union(true, union.getInputs().size());
+
+    // Create the top aggregate. We must adjust group key indexes of the
+    // original aggregate. E.g., if the original tree was:
+    //
+    // Aggregate[groupSet=$1, ...]
+    //   Union[...]
+    //
+    // Then the new tree should be:
+    // Aggregate[groupSet=$0, ...]
+    //   Union[...]
+    //     Aggregate[groupSet=$1, ...]
+    Mapping topGroupMapping = Mappings.create(MappingType.INVERSE_SURJECTION,
+        union.getRowType().getFieldCount(),
+        aggRel.getGroupCount()
+    );
+    int i = 0;
+    for (int groupKey : aggRel.getGroupSet()) {
+      topGroupMapping.set(groupKey, i++);
+    }
+
+    ImmutableBitSet topGroupSet = Mappings.apply(topGroupMapping, aggRel.getGroupSet());
+    ImmutableList.Builder<ImmutableBitSet> builder = ImmutableList.builder();
+    for (ImmutableBitSet groupSet : aggRel.getGroupSets()) {
+      builder.add(Mappings.apply(topGroupMapping, groupSet));
+    }
+    ImmutableList<ImmutableBitSet> topGroupSets = builder.build();
```

Comment:
Consider using `Mappings.apply2(Mapping, Iterable<ImmutableBitSet>)`; see how that method is used in `AggregateJoinTransposeRule`.

Reply:
Fixed in the latest commit: https://github.com/apache/calcite/blob/97e4483ccdf9232ae1f879c7b05b7629708712d3/core/src/main/java/org/apache/calcite/rel/rules/AggregateUnionTransposeRule.java#L177-L178